hack/api-docs/build.sh

```diff
@@ -9,31 +9,43 @@
 set -euo pipefail
 
 SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+REPO_ROOT="${SCRIPT_DIR}/../.."
 DOCS_DIR="${SCRIPT_DIR}/../../docs"
-REFDOCS_PKG="github.com/elastic/gen-crd-api-reference-docs"
+REFDOCS_REPO="https://github.com/elastic/gen-crd-api-reference-docs.git"
 REFDOCS_VER="master"
-REFDOCS_BIN="$(go env GOPATH)/bin/$(basename $REFDOCS_PKG)"
 
-install_refdocs() {
+log() {
+    echo ">> $1"
+}
+
+build_docs() {
     local INSTALL_DIR=$(mktemp -d)
     (
+        log "Building refdoc binary..."
         cd $INSTALL_DIR
-        GO111MODULE=on go mod init github.com/elastic/eck-refdocs
-        GO111MODULE=on go get -u "${REFDOCS_PKG}@${REFDOCS_VER}"
+        git clone -q --depth 1 --branch $REFDOCS_VER --single-branch $REFDOCS_REPO
```

Comment:
Could we not cut a new release of our fork and `go get github.com/elastic/gen-crd-api-reference-docs@0.1.6`?

Reply:
Just `go get` won't work any more because the `go.mod` file needs to be modified to point to the current state of the operator source.

Reply:
Can you clarify what you mean here? I think it is failing because the last tag in our fork still had the upstream module name: https://github.com/elastic/gen-crd-api-reference-docs/blob/v0.1.5/go.mod

but that is changed in master, so I thought cutting a new release would fix it

Reply:
The main reason the generation step started to fail after the upgrade to Kubebuilder v2 is because gengo wasn't module aware (it still uses the underlying Go tooling but some things don't translate well), and because it was specifically looking for annotations like `+genclient` and group names being present in `doc.go` files. 

For being module aware, I believe the community is in the process of updating gengo tooling but it is a long process. One trick to make the existing code work is by adding a `replace` directive that points the module query result for `latest` to the source tree. This is what I am doing here in this shell script. I could of course do a `go get`, switch directory to `$GOPATH` and do this bit, but I didn't want to pollute the GOPATH of someone who might want to run this script locally. 

I haven't tagged the fork yet because little issues keep cropping up and it's easier to just work off of master until everything settles down. I am planning to tag the release after this PR is merged.

Reply:
Summing up some out of band conversation:
- `go get github.com/elastic/gen-crd-api-reference-docs` fails currently because we do not have a tagged release of our fork with the go.mod declaring itself as the fork, so you receive
```
go get: github.com/elastic/gen-crd-api-reference-docs@v0.1.5: parsing go.mod:
    module declares its path as: github.com/ahmetb/gen-crd-api-reference-docs
            but was required as: github.com/elastic/gen-crd-api-reference-docs
```
we could resolve this by either tagging a new release or `go get @master`, BUT that is not the only issue

- gengo (used by the generator) also wants to pull down the `latest` package of cloud-on-k8s, but we want to operate against the local file system, which is why we need to add the replace directive to the go.mod in the generator to use the local filesystem. This should be fixed in a later version of gengo. I'm good to merge this if we add the upstream gengo issue as a TODO either in the code or as a github issue in our repo with a short explanation

Reply:
Thanks for the summation. To be clear, the first point with `go get` was never an issue. Prior to the Kubebuilder + Go modules upgrade, the script [simply used to `go get github.com/elastic/gen-crd-api-reference@master`](https://github.com/elastic/cloud-on-k8s/blob/c4212c63cf31864cb63ad1ea1112ff7b2085398b/hack/api-docs/build.sh#L22).  After the upgrade to Go modules, the download still worked but `gengo` project's parsing behaviour changed and it was querying for the `latest` version of the `github.com/elastic/cloud-on-k8s` -- which fails (reason unclear). The workaround is to add a `replace` directive to the `go.mod` of `gen-crd-api-reference`. Doing this on the source checked out to GOPATH via `go get` is not ideal -- hence why the script does a `git clone` instead.

Even though adding the `replace` directive is a workaround in response the changed behaviour of `gengo`, I think it is the correct thing to do anyway because we always want to scan the _current_ copy of the ECK source and not what's on GitHub or the Go module cache.

The `gengo` issue mentioned here is https://github.com/kubernetes/gengo/issues/147. I wrote some parsing code myself using `x/tools/package` to scan the ECK source and it didn't require the `replace` directive to work, so I am assuming that once that issue is resolved, the "workaround" probably won't be necessary. However, as stated above, explicitly pointing to the current source tree _is_ the behaviour we want, so I don't think it's critical to closely follow the upstream issue as it is not a bug but a feature IMO :smile: